# Prepare data

## Load gut data

Load the original data files outputted by the bioinformatic pipeline.

### Sample metadata

```{r load_sample_metadata_gut, warning=FALSE, comments="", message=FALSE, eval=FALSE}
sample_metadata <- read_csv2("data/squirrels_metadata.csv") %>% 
  arrange(sample)
```

### Genome metadata

```{r load_genome_metadata, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_metadata <- read_csv("data/squirrels_mag_info.csv") %>% 
  rename(length=mag_size) %>% 
  arrange(genome)
```

### Sequencing reads

```{r load_read_counts, warning=FALSE, comments="", message=FALSE, eval=FALSE}
read_counts <- read_csv2("data/squirrels_counts.csv") %>%
    rename(genome=1) %>%
    arrange(match(genome, genome_metadata$genome)) %>%
    select(c("genome",sample_metadata$sample))
```

```{r load_genome_coverage, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_coverage <- read_csv2("data/squirrels_coverage.csv") %>%
    rename(genome=1) %>%
    arrange(match(genome, genome_metadata$genome)) %>%
    select(c("genome",sample_metadata$sample))
```

### Genome tree

```{r load_genome_tree_gut, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_tree <- read_tree("data/squirrels.tree")
genome_tree$tip.label <- str_replace_all(genome_tree$tip.label,"'", "") #remove single quotes in MAG names
genome_tree <- keep.tip(genome_tree, tip=genome_metadata$genome) # keep only MAG tips
```

### Genome annotations

```{r load_genome_annotations, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_annotations <- read_tsv("data/genome_annotations.tsv.xz") %>%
    rename(gene=1, genome=2, contig=3)
```

## Create working objects

Transform the original data files into working objects for downstream analyses.

### Filter reads by coverage

```{r filter_coverage, warning=FALSE, comments="", message=FALSE, eval=FALSE}
min_coverage=0.3
read_counts_filt <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]]))
```

### Transform reads into genome counts

```{r calculate_genome_counts_unfiltered, warning=FALSE, comments="", message=FALSE, eval=FALSE}
readlength=150
genome_counts <- read_counts %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

```{r calculate_genome_counts_filtered, warning=FALSE, comments="", message=FALSE, eval=FALSE}
readlength=150
genome_counts_filt <- read_counts_filt %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

### Distill annotations into GIFTs

```{r distill_annotations_gut, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_gifts <- distill(genome_annotations,GIFT_db,genomecol=2,annotcol=c(9,10,19), verbosity=F)
```

## Prepare color scheme

[AlberdiLab](www.alberdilab.dk) projects use unified color schemes developed for the [Earth Hologenome Initiative](www.earthhologenome.org), to facilitate figure interpretation.

```{r get_ehi_colors_gut, warning=FALSE, comments="", message=FALSE, eval=FALSE}
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>%
    unique() %>%
    arrange(phylum) %>%
    pull(colors, name=phylum)
```

```{r get_colors_gut, warning=FALSE, comments="", message=FALSE, eval=FALSE}
altitude_colors<- c("#429ef5","#a8c3f0","#f56c42","#f29c80")
treatment_colors <- c("#429ef5","#f56042")
```

## Wrap working objects

All working objects are wrapped into a single Rdata object to facilitate downstream usage.

```{r wrap_working_objects_gut, warning=FALSE, comments="", message=FALSE, eval=FALSE}
save(sample_metadata,
     genome_metadata,
     read_counts,
     genome_counts,
     genome_counts_filt,
     genome_tree,
     genome_gifts,
     phylum_colors,
     altitude_colors,
     treatment_colors,
     file = "data/gut/data.Rdata")
```

## Load skin data

Load the original data files outputted by the bioinformatic pipeline.

### Reads
```{r load_reads_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
read_counts <- read_tsv("data/skin/DMB0171_counts.tsv.gz") %>%
    rename(genome=1)
```

### Sample metadata
```{r load_sample_metadata_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
sample_metadata <- read_csv("data/sample_metadata.csv") %>%
  select("EHI_number","capture_id_flat","sample_type", "sex", "latitude", "longitude","singlem_fraction","reads_post_fastp","metagenomic_bases","host_bases", "bases_lost_fastp_percent_flat") %>%
  rename(sample=1) %>%
  filter(sample_type %in% c("Skin swab: dorsal", "Skin swab: ventral")) %>%
  filter(sample %in% colnames(read_counts)) %>%
  separate(sample_type, c("sample_type", "position"), sep = ":") %>%
  mutate(environment = case_when(
        longitude < -1.5 ~ "low",
        longitude > -1.5 ~ "high",
        TRUE ~ NA_character_ # Default case if no match
      ))%>%
  mutate(river = case_when(
        latitude == 43.00 ~ "Erlan",
        latitude == 43.03 ~ "Harpea",
        latitude == 43.16 ~ "Goizueta",
        latitude == 43.14 ~ "Leitzaran",
        TRUE ~ NA_character_ # Default case if no match
      ))%>%
  select(, -sample_type, -latitude, -longitude)
```

```{r}
#  filter(!sample %in% c("EHI02973", "EHI02983", "EHI02978", "EHI02987", "EHI02976", "EHI02988", ))%>%
```

### Read counts arrange

```{r load_read_counts_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
read_counts <- read_counts %>%
    select(one_of(c("genome",sample_metadata$sample)))
```

### Genome taxonomy

```{r load_genome_metadata_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_metadata <- read_tsv("data/skin/DMB0171_mag_info.tsv.gz") %>%
  rename(length=mag_size)%>%
  semi_join(., read_counts, by = "genome") %>%
  arrange(match(genome,read_counts$genome))
```

### Genome base hits

```{r load_genome_hits_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_coverage <- read_tsv("data/skin/DMB0171_coverage.tsv.gz") %>%
  rename(genome=1) %>%
  select(one_of(c("genome",sample_metadata$sample)))%>%
  semi_join(., genome_metadata, by = "genome")%>%
  arrange(match(genome, read_counts$genome))
```

### Genome tree

```{r load_genome_tree_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_tree <- read_tree("data/skin/DMB0171.tree")
genome_tree$tip.label <- str_replace_all(genome_tree$tip.label,"'", "") #remove single quotes in MAG names
genome_tree <- keep.tip(genome_tree, tip=genome_metadata$genome) # keep only MAG tips
```

### Genome annotations

Downloading individual annotation files from ERDA using information in Airtable and writing them to a single compressed table takes a while. The following chunk only needs to be run once, to generate the ***genome_annotations*** table that is saved in the data directory. Note that the airtable connection requires a personal access token.

```{r download_genome_annotations_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
airtable("MAGs", "appWbHBNLE6iAsMRV") %>% #get base ID from Airtable browser URL
  read_airtable(., fields = c("ID","mag_name","number_genes","anno_url"), id_to_col = TRUE) %>% #get 3 columns from MAGs table
  filter(mag_name %in% paste0(genome_metadata$genome,".fa")) %>% #filter by MAG name
  filter(number_genes > 0) %>% #genes need to exist
  select(anno_url) %>% #list MAG annotation urls
  pull() %>%
  read_tsv() %>% #load all tables
  rename(gene=1, genome=2, contig=3) %>% #rename first 3 columns
  write_tsv(file="data/skin/genome_annotations.tsv.xz") #write to overall compressed file
```

```{r load_genome_annotations_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_annotations <- read_tsv("data/skin/genome_annotations.tsv.xz") %>%
    rename(gene=1, genome=2, contig=3)
```

## Create working objects

Transform the original data files into working objects for downstream analyses.

### Filter reads by coverage

```{r filter_coverage_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
min_coverage=0.3
read_counts_filt <- genome_coverage %>%
  mutate(across(where(is.numeric), ~ ifelse(. > min_coverage, 1, 0))) %>%
  mutate(across(-1, ~ . * read_counts[[cur_column()]]))
```

### Transform reads into genome counts

```{r calculate_genome_counts_unfiltered_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
readlength=150
genome_counts <- read_counts %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

```{r calculate_genome_counts_filtered_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
readlength=150
genome_counts_filt <- read_counts_filt %>%
  mutate(across(where(is.numeric), ~ . / (genome_metadata$length / readlength) ))
```

```{r filter_lost_samples_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
sample_metadata <- sample_metadata %>%
  filter(sample %in% colnames(genome_counts_filt))
```

### Distill annotations into GIFTs

```{r distill_annotations_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
genome_gifts <- distill(genome_annotations,GIFT_db,genomecol=2,annotcol=c(9,10,19), verbosity=F)
```

## Prepare color scheme

[AlberdiLab](www.alberdilab.dk) projects use unified color schemes developed for the [Earth Hologenome Initiative](www.earthhologenome.org), to facilitate figure interpretation.

```{r get_ehi_colors_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
phylum_colors <- read_tsv("https://raw.githubusercontent.com/earthhologenome/EHI_taxonomy_colour/main/ehi_phylum_colors.tsv") %>%
    right_join(genome_metadata, by=join_by(phylum == phylum)) %>%
    arrange(match(genome, genome_tree$tip.label)) %>%
    select(phylum, colors) %>%
    unique() %>%
    arrange(phylum) %>%
    pull(colors, name=phylum)
```

```{r get_colors_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
altitude_colors<- c("#429ef5","#a8c3f0","#f56c42","#f29c80")
treatment_colors <- c("#429ef5","#f56042")
```

## Wrap working objects

All working objects are wrapped into a single Rdata object to facilitate downstream usage.

```{r wrap_working_objects_skin, warning=FALSE, comments="", message=FALSE, eval=FALSE}
save(sample_metadata,
     genome_metadata,
     read_counts,
     genome_counts,
     genome_counts_filt,
     genome_tree,
     genome_gifts,
     phylum_colors,
     altitude_colors,
     treatment_colors,
     file = "data/skin/data.Rdata")
```
